### PR TITLE
Windows - Visual Studio port

### PR DESCRIPTION
**Pull request by fsrajer**
Port for Windows - Visual Studio. Tested on Windows 7 and 10 with Visual Studio 2013 Update 5. Implemented in the style
#ifdef WIN32
so that the changes are applied only on Windows. This was verified and is building on Ubuntu 16.04.

### CODE REVIEW

File: Core/src/Utils/DeformationGraph.cpp [hunk 1]
```diff
@@ -169,13 +174,13 @@ void DeformationGraph::setPosesSeq(std::vector<unsigned long long int> * poseTim
 
         imin = std::min(imin, (int)sampledGraphTimes.size() - 1);
 
-        if(abs(sampledGraphTimes[imin] - poseTime) <= abs(sampledGraphTimes[imid] - poseTime) &&
-           abs(sampledGraphTimes[imin] - poseTime) <= abs(sampledGraphTimes[imax] - poseTime))
+        if(ulonglongDiffAbs(sampledGraphTimes[imin],poseTime) <= ulonglongDiffAbs(sampledGraphTimes[imid],poseTime) &&
```
Comment (mp3guy) [inline Core/src/Utils/DeformationGraph.cpp]: Why not just use std::abs?

File: Core/src/Utils/WindowsExtras.cpp [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+
+#ifdef WIN32
+
+#include <stdint.h>
+#include <string.h>
+#define far
+#include <Windows.h>
+#include <WinSock2.h>
+
+int gettimeofday(struct timeval * tp,struct timezone * tzp)
```
Comment (mp3guy) [inline Core/src/Utils/WindowsExtras.cpp]: I would guess we can just use std::chrono anywhere this is required instead?

File: GUI/src/CMakeLists.txt [hunk 3]
```diff
@@ -4,12 +4,18 @@ project(ElasticFusion)
 
 set(CMAKE_MODULE_PATH ${CMAKE_MODULE_PATH} "${CMAKE_CURRENT_SOURCE_DIR}")
 
+if (WIN32)
+  find_package(JPEG REQUIRED)
```
Comment (mp3guy) [inline GUI/src/CMakeLists.txt]: Surely this gets added as a requirement when you build Pangolin, and hence shouldn't be needed?

File: GUI/src/Findefusion.cmake [hunk 4]
```diff
@@ -25,4 +25,4 @@ set(EFUSION_LIBRARIES ${EFUSION_LIBRARY})
 
 include(FindPackageHandleStandardArgs)
 find_package_handle_standard_args(EFUSION DEFAULT_MSG EFUSION_LIBRARY EFUSION_INCLUDE_DIR)
-mark_as_advanced(EFUSION_LIBRARY EFUSION_INCLUDE_DIR)
+#mark_as_advanced(EFUSION_LIBRARY EFUSION_INCLUDE_DIR)
```
Comment (mp3guy) [inline GUI/src/Findefusion.cmake]: Why is this?

File: README.md [hunk 5]
```diff
@@ -1,163 +1,183 @@
...
+* [OpenNI2](https://github.com/occipital/OpenNI2)
+* [SuiteSparse] (https://github.com/jlblancoc/suitesparse-metis-for-windows)
+* [Eigen] (http://eigen.tuxfamily.org)
+* [Pangolin](https://github.com/stevenlovegrove/Pangolin)
+  * zlib (Pangolin can automatically download and build this)
+  * libjpeg (Pangolin can automatically download and build this)
+
+Firstly install cmake and cuda. Then download and build from source OpenNI2, SuiteSparse. Next download Eigen (no need to build it since it is a header-only library). Then download and build from source Pangolin but pay attention to the following cmake settings. There will be a lot of dependencies where path was not found. That is OK except OPENNI2 and EIGEN3 (those should be set to valid paths). You also need to set MSVC_USE_STATIC_CRT to false in order to correctly link to ElasticFusion projects. Also, you can set BUILD_EXAMPLES to false since we don't need them and some were crashing on my machine.
+
+Next, build Core from source. Note that the build might crash if you are missing dirent.h (If so, look here https://github.com/tronkko/dirent). Finally, build GUI.
```
Comment (mp3guy) [inline README.md]: So is there a dependency on dirent.h or not? If so, that should be removed if possible

File: GUI/src/Findefusion.cmake (hunk 4 above)
Comment (fsrajer) [inline GUI/src/Findefusion.cmake]: This is only for convenience. cmake on windows is often unable to find paths to libraries and you have to set it manually. This is true also for efusion.lib. If you use cmake-gui, then these two variables get mixed up with other cmake advanced variables and are hidden in the list by default.

File: Core/src/Utils/DeformationGraph.cpp (hunk 1 above)
Comment (mp3guy) [inline Core/src/Utils/DeformationGraph.cpp]: Sure, but maybe it should be cast to int64_t just before to simplify the logic.

File: README.md (hunk 5 above)
Comment (fsrajer) [inline README.md]: Sorry for this. It seems that dirent.h could have been simply omitted on Windows. I have commited the change now.

File: Core/src/Utils/DeformationGraph.cpp (hunk 1 above)
Comment (mp3guy) [inline Core/src/Utils/DeformationGraph.cpp]: Time comes from tick which starts at 1; https://github.com/mp3guy/ElasticFusion/blob/master/Core/src/ElasticFusion.cpp#L52. So we can safely use option 1.

Comment (fsrajer) [inline Core/src/Utils/DeformationGraph.cpp]: OK, fixed.

File: README.md (hunk 5 above)
Comment (zhangtengyu77) [inline README.md]: I have compiled your program on Windows 10, but every time the program crushed on the function "glCreateProgram" in pangonlin, and I'm sure the pangolin is well installed and tested.

**Comment by M1234Thomas** [comment]
Hello,

I am using the 64 bit version of visual studio an I am the getting the below error.
error LNK2019: unresolved external symbol "__declspec(dllimport) and its tracing back to the ifdef win 32.

Is the above code supposed to work in 64 bit environment as well?

**Comment by fsrajer** [comment]
Yes, it should work with 64. Actually, I haven't even tested it with 32. So the problem will be elsewhere.

That sounds like that mempcpy and gettimeofday are not getting exported. Is WIN32 and efusion_EXPORTS defined in efusion project in "project settings -> c/c++ -> preprocessor"?

**Comment by M1234Thomas** [comment]
Thanks for the response. The efusion_EXPORTS preprocessor definition was missing. It works now :)

**Comment by fsrajer** [comment]
Good to hear that! CMake should define this automatically though. If you figure out why it did not happen automatically for you, you could post the patch.

**Pull request merged**